posts/Gunnar/2018-03-19-spec-api-modularity-patterns.adoc

```diff
@@ -0,0 +1,426 @@
+= "Spec API Modularity Patterns"
+Gunnar Morling
+:awestruct-tags: [ "Discussions" ]
+:awestruct-layout: blog-post
+---
+
+While Java 9 http://www.oracle.com/technetwork/java/eol-135779.html[reaches its end-of-life] this month,
+the Java Platform Module System (JPMS, https://www.jcp.org/en/jsr/detail?id=376[JSR 376]) is here to stay.
+That means that also specifications such as the https://www.jcp.org/en/jsr/detail?id=338[Java Persistence API] or https://www.jcp.org/en/jsr/detail?id=380[Bean Validation]
+will eventually have to be adjusted to support and take advantage of the module system.
+
+This blog post is planned to be the first of a series on exploring JPMS modularity patterns for specification APIs.
+In this part we're going to look into how implementations of a specification API can be bootstrapped in a portable way
+and how such implementations can access the private state of modules of the user of the API.
+link:/2017/04/11/accessing-private-state-of-java-9-modules/[As discussed a while ago],
+the latter is a common requirement;
+for instance JPA providers such as Hibernate ORM must do so for reading and writing entity state.
+
+As an example, let's consider three JPMS modules:
+
+* _fieldreader.spec_: a made-up "specification" which defines a single contract for accessing private fields of user classes
```

Comment:
Remove "single" ? it's irrelevant - but I'm not sure it's irrelevant- so it becomes confusing